src/Microsoft.AspNetCore.Razor/Parser/TagHelpers/TagHelperParseTreeRewriter.cs

```diff
@@ -307,32 +315,98 @@ private IEnumerable<string> GetAttributeNames(Block tagBlock)
 
             if (childCount <= 1)
             {
-                return Enumerable.Empty<string>();
+                return Enumerable.Empty<KeyValuePair<string, string>>();
             }
 
-            var attributeChildren = new List<SyntaxTreeNode>(childCount - 1);
-            for (var i = 1; i < childCount; i++)
-            {
-                attributeChildren.Add(tagBlock.Children[i]);
-            }
-            var attributeNames = new List<string>();
+            _htmlAttributeTracker.Clear();
 
-            foreach (var child in attributeChildren)
+            var attributes = _htmlAttributeTracker;
+
+            for (var i = 1; i < childCount; i++)
             {
+                var child = tagBlock.Children[i];
                 Span childSpan;
 
                 if (child.IsBlock)
                 {
-                    childSpan = ((Block)child).FindFirstDescendentSpan();
+                    var childBlock = (Block)child;
+                    childSpan = childBlock.FindFirstDescendentSpan();
 
                     if (childSpan == null)
                     {
+                        _attributeValueBuilder.Append(InvalidAttributeValueMarker);
                         continue;
                     }
+
+                    var childOffset = 0;
```

Comment:
`childOffset` calculation seems to assume the first `Span` always contains all of `attributeName="` or `attributeName='`. Is that an invariant even when there's whitespace before or after the `=`?

Then the use of `childOffset` below relies on the final `Span` containing at most the final quote or double quote and any trailing whitespace. Is this also an invariant?